### PR TITLE
Update to cslib_object

### PR DESCRIPTION
**Pull request by StephenXHart**
Added a subroutine get_by_field which allows an object to be retrieved via an arbitrary field matching an arbitrary value.

### CODE REVIEW

File: scenes/cslib_object.txt [hunk 1]
```diff
@@ -70,6 +79,46 @@
 *gosub _set_main_instance p_objectname p_index
 *return
 
+*comment GET_BY_FIELD
+*comment ------------------------------------------------------------------
+*comment Takes the specified field name and field value and finds the
+*comment first instance of a match that it findscl and copies that data
```
Comment (diegoboston) [inline scenes/cslib_object.txt]: findscl ?

File: scenes/cslib_object.txt [hunk 2]
```diff
@@ -70,6 +79,46 @@
...
+*comment returns:
+*comment	none
+*comment usage example:
+*comment 	*gosub_scene cslib_object get "dragon" "name" "Wally"
+*comment ------------------------------------------------------------------
+
+*label get_by_field
+*params p_object p_field p_value
+*set cslib_ret ""
+*temp foundit false
```
Comment (diegoboston) [inline scenes/cslib_object.txt]: maybe found_it ?

**Review by diegoboston — APPROVED** [review]
THanks!

File: scenes/_test_entry.txt [hunk 3]
```diff
@@ -19,6 +19,7 @@
 		*gosub_scene _test_display_templates display_module_summary module[module_id] tests_run tests_passed
 		Results have been logged.
 		*page_break
+		*gosub view_logs
```
Comment (CareyJWilliams) [inline scenes/_test_entry.txt]: This looks unrelated?

File: scenes/_test_entry.txt [hunk 4]
```diff
@@ -76,4 +77,4 @@ Results have been logged.
 	*gosub_scene _test_display_templates display_log cslib_ret
 	*page_break
 	*goto log_selection
-*return
\ No newline at end of file
+*return
```
Comment (CareyJWilliams) [inline scenes/_test_entry.txt]: And this?

File: scenes/cslib_object.txt (hunk 1 above)
Comment (CareyJWilliams) [reply scenes/cslib_object.txt]: Yep, please fix this!

File: scenes/cslib_object.txt (hunk 2 above)
Comment (CareyJWilliams) [reply scenes/cslib_object.txt]: As Diego says :)

File: scenes/cslib_object.txt [hunk 5]
```diff
@@ -70,6 +79,46 @@
...
+*temp n 1
+*temp array_count {p_object&"_max"}
```
Comment (CareyJWilliams) [inline scenes/cslib_object.txt]: By committing this we are agreeing to hold objects to array standards in cslib.
I'm fine with this, but @diegoboston, weren't you hesitant on the last PR?

File: scenes/startup.txt [hunk 6]
```diff
@@ -123,4 +126,4 @@
 		All tests passed.
 *else
 	*comment allow user to select individual tests, view logs etc.
-	*goto_scene _test_entry main
\ No newline at end of file
+	*goto_scene _test_entry main
```
Comment (CareyJWilliams) [inline scenes/startup.txt]: Something odd is going on with the line endings I think?

File: scenes/startup.txt [hunk 7]
```diff
@@ -102,6 +102,9 @@
 *create dragon_1_hit_points 100
 *create dragon_2_name "Hobbes"
 *create dragon_2_hit_points 150
+*create dragon_3_name "Stebbing"
+*create dragon_3_hit_points 100
+*create dragon_max 3
```
Comment (CareyJWilliams) [inline scenes/startup.txt]: If we're going to use _max, we should define _count as well (see: https://github.com/ChoicescriptIDE/cslib/wiki/Arrays), for consistency (even if it's not used).

File: scenes/cslib_object.txt (hunk 5 above)
Comment (CareyJWilliams) [reply scenes/cslib_object.txt]: We should probably use _count instead of _max here (see: https://github.com/ChoicescriptIDE/cslib/wiki/Arrays).

**Review by CareyJWilliams — CHANGES_REQUESTED** [review]

File: scenes/cslib_object.txt [hunk 8]
```diff
@@ -110,3 +159,19 @@
...
+*comment --- ------------------------------------------------------------------
+*label _clear_main_instance
+*params p_object
+*temp n 1
+*temp field_count {p_object&"_field_max"}
+*temp field ""
+*label _clear_main_loop
+*if (n <= field_count)
+	*set field {p_object&"_field_${n}"}
+	*set {p_object&"_${field}"} false
```
Comment (CareyJWilliams) [inline scenes/cslib_object.txt]: I think we should use "" rather than _false_ here, to keep it consistent with resetting cslib_ret. It's not perfect, but it's the closest we have to null/undefined. _false_ is more likely to clash with an intended value (e.g. dragon_alive: false).

File: scenes/test_object.txt [hunk 9]
```diff
@@ -13,21 +13,37 @@
...
+*gosub test_finish
+
 *gosub test_start "set (modify) and get (modified) data"
 *gosub_scene cslib_object get "dragon" 2
 *gosub test_assert (dragon_2_name != "Locke") "=> Expected - anything but Locke[n/]=> Received - ${dragon_2_name}[n/]"
 *gosub_scene cslib_object set "dragon" 2 "name" "Locke"
 *gosub_scene cslib_object get "dragon" 2
 *gosub test_assert (dragon_2_name = "Locke") "=> Expected - Locke[n/]=> Received - ${dragon_2_name}[n/]"
 *gosub test_assert (dragon_name = dragon_2_name) "=> Expected - Locke[n/]=> Received - ${dragon_name}[n/]"
+*gosub_scene cslib_object get_by_field "dragon" "name" "Locke"
```
Comment (CareyJWilliams) [inline scenes/test_object.txt:55]: Very diligent! :)

File: scenes/cslib_object.txt [hunk 10]
```diff
@@ -36,12 +39,18 @@
 *comment *create dragon_1_hp 		100
 *comment *create dragon_2_name 		"Dave"
 *comment *create dragon_2_hp		150
+*comment *create dragon_max			2
```
Comment (CareyJWilliams) [inline scenes/cslib_object.txt]: Check the indentation here: spaces vs tabs?

**Review by CareyJWilliams — CHANGES_REQUESTED** [review]

File: scenes/_test_entry.txt (hunk 3 above)
Comment (StephenXHart) [reply scenes/_test_entry.txt]: Oops. Some test stuff. I'll remove it.

File: scenes/cslib_object.txt (hunk 1 above)
Comment (StephenXHart) [reply scenes/cslib_object.txt]: covfefe. No idea where that came from.

File: scenes/cslib_object.txt (hunk 2 above)
Comment (StephenXHart) [reply scenes/cslib_object.txt]: OK

File: scenes/cslib_object.txt (hunk 5 above)
Comment (StephenXHart) [reply scenes/cslib_object.txt]: OK

File: scenes/cslib_object.txt (hunk 8 above)
Comment (StephenXHart) [reply scenes/cslib_object.txt]: The reason I chose false rather than "" is that if you have a boolean field within your object (dragon_alive) and you set it to "" then the statement:
*if (dragon_alive)
    *goto xxx
will barf with a Neither true nor false error.
While false might cause confusion it won't actually cause a crash.

File: scenes/_test_entry.txt (hunk 4 above)
Comment (StephenXHart) [reply scenes/_test_entry.txt]: Don't understand this. The old and new statements look identical?

File: scenes/startup.txt (hunk 6 above)
Comment (StephenXHart) [reply scenes/startup.txt]: This seems to be the whole Windows (CRLF) vs Unix (LF) thing. I work on a Mac (which is of course really Unix).
I haven't worked out a way of dealing with this yet.

Comment (StephenXHart) [reply scenes/startup.txt]: I have configured git so that any LF endings will be converted to CRLF on commit but there still seem to be a few problems.

File: scenes/cslib_object.txt (hunk 10 above)
Comment (StephenXHart) [reply scenes/cslib_object.txt]: OK

**Comment by StephenXHart** [comment]
I've provided an update, but randomtest is failing with:

RANDOMTEST FAILED: Error: cslib_object line 106: Non-existent variable 'dragon_count'

dragon_count only appears in a comment field so I'm not sure what's happening.

EDIT: ignore this comment; I failed to update the test data (doh)

File: scenes/startup.txt (hunk 6 above)
Comment (CareyJWilliams) [reply scenes/startup.txt]: No worries, I'll push a fix before we merge it :)

File: scenes/cslib_object.txt (hunk 8 above)
Comment (CareyJWilliams) [reply scenes/cslib_object.txt]: I think I'd prefer the crash: explicit errors are much easier to diagnose than logical ones. What do you think @diegoboston ?

Comment (CareyJWilliams) [reply scenes/cslib_object.txt]: And @danspinola ?

File: scenes/startup.txt [hunk 11]
```diff
@@ -104,7 +104,7 @@
 *create dragon_2_hit_points 150
 *create dragon_3_name "Stebbing"
 *create dragon_3_hit_points 100
-*create dragon_max 3
```
Comment (CareyJWilliams) [inline scenes/startup.txt]: Please leave _max defined as well

File: scenes/_test_entry.txt (hunk 4 above)
Comment (CareyJWilliams) [reply scenes/_test_entry.txt]: It's probably line-ending related, I'll tidy it up before we merge.

File: scenes/cslib_object.txt (hunk 8 above)
Comment (StephenXHart) [reply scenes/cslib_object.txt]: I don't think using a library should cause (perfectly legal) code to fail.

It might put people off using it apart from anything else, and that would be a waste. :)

File: scenes/startup.txt (hunk 11 above)
Comment (StephenXHart) [reply scenes/startup.txt]: Happy to do so, but why do we need both?

File: scenes/cslib_object.txt (hunk 8 above)
Comment (CareyJWilliams) [reply scenes/cslib_object.txt]: I appreciate what you're saying, but that argument doesn't hold any better for other examples:

```
*gosub_scene cslib_object get_by_field "dragon" "name" "idonotexist"
*if (dragon_hit_points > 0)
=> ChoiceScript Error: false is not a number
  
```

Whatever we set it to the user is going to have to do their own sanity checks, or risk crashes, regardless of what value we "reset" attributes to.

And between "" and _false_, the former has the slight advantages of:
- being much less likely to be used as a meaningful value (a string holding a name is much less likely to be empty than a boolean value is to be false)
- consistency with cslib_ret reset

An alternative option is to just throw a `*bug cslib_object.get_by_field: no matching object found` error rather than clearing the main instance? There is no reliable way to confirm whether or not a valid object was returned at the moment. There would be if we had an _id field on the main instance (which I still think is a good idea generally). Then you could set the _id field to a negative number here, and expect users to check it's a positive value before accessing attributes.

Comment (StephenXHart) [reply scenes/cslib_object.txt]: It's pretty much an intractable problem.

Unless we require people to define field types to which I say 'no - what on earth are you thinking?!

We could, in theory, provide a configuration option but that would involve another cslib global which wouldn't be a good idea either.

I'm still in favour of _false_, but I'll go with the majority..

Forcing an _id field is kind of a separate issue although I'm not against it. Setting it to a negative (or whatever) value is useful but you still have the question of what to do with the rest of the fields - do you clear them or leave them with their current (wrong) values?

Comment (StephenXHart) [reply scenes/cslib_object.txt]: This may ultimately not matter all that much.

I find that developing with ChoiceScript basically involves stopping it crashing. If I fail to find dragon 'Fnord' and it barfs I go back and look at it until I realise that I made a typo and I meant 'Ford'.

Checking return values is seldom worth the effort. (I know, I know. Bad programming style and all that but it seems to be what happens)

Comment (StephenXHart) [reply scenes/cslib_object.txt]: Yet another alternative would be to use Unix-style error codes and messages.

Required fields:
*create dragon_err
*create dragon_errmsg

Where dragon_err will be 0 if there is no error and dragon_errmsg is optional.

This might be overkill.

Comment (CareyJWilliams) [reply scenes/cslib_object.txt]: It sounds to me like you might be in favour of throwing a *bug instead of clearing on a bad search value? Shall we go with that? As you say, it is quite ChoiceScriptish. But, more importantly, it means we don't have to come to any more controversial conclusion of what to do with the fields. We can punt this problem down the road.

Comment (StephenXHart) [reply scenes/cslib_object.txt]: Works for me. At least it's equally bad for everyone...

**Comment by StephenXHart** [comment]
Changed get_by_field failure from false to *bug.
I've also kept the subroutine _clear_main_instance even though we aren't using it - I gave it a parameter to which to set all fields. Leaving it there as potentially useful later.
I think I got the tests right for handling *bug but please check.
Haven't added [dragon]_max - why do we need it?

**Comment by CareyJWilliams** [comment]
> Changed get_by_field failure from false to *bug.
> I've also kept the subroutine _clear_main_instance even though we aren't using it - I gave it a parameter to which to set all fields. Leaving it there as potentially useful later.
> I think I got the tests right for handling *bug but please check.
> Haven't added [dragon]_max - why do we need it?

LGTM: You'd have got test failures if you'd messed up the *bug test, so great job :)
I'll do a final pass now.

File: scenes/test_object.txt [hunk 12]
```diff
@@ -13,21 +13,39 @@
...
+*gosub test_assert (dragon_name = dragon_2_name) "=> Expected - ${dragon_2_name}[n/]=> Received - ${dragon_name}[n/]"
+*gosub_scene cslib_object get_by_field "dragon" "hit_points" 100 
+*gosub test_assert (dragon_name = dragon_1_name) "=> Expected - ${dragon_1_name}[n/]=> Received - ${dragon_name}[n/]"
+*gosub_scene cslib_object get_by_field "dragon" "hit_points" 150 
+*gosub test_assert (dragon_name = dragon_2_name) "=> Expected - ${dragon_2_name}[n/]=> Received - ${dragon_name}[n/]"
+*set cslib_catch_bug true
+*gosub_scene cslib_object get_by_field "dragon" "name" "Unknown" 
+*set cslib_catch_bug false
+*gosub test_assert_equal "Bug: cslib_object.get_by_field: Value of 'Unknown' not found for field 'name' in object type 'dragon'" cslib_bug_message
+*gosub test_finish
```
Comment (CareyJWilliams) [inline scenes/test_object.txt:45]: Let's move this into its own test:
```suggestion
*gosub test_finish

*gosub test_start "throw a bug on no match"
*set cslib_catch_bug true
*gosub_scene cslib_object get_by_field "dragon" "name" "Unknown" 
*set cslib_catch_bug false
*gosub test_assert_equal "Bug: cslib_object.get_by_field: Value of 'Unknown' not found for field 'name' in object type 'dragon'" cslib_bug_message
*gosub test_finish
```

File: scenes/cslib_object.txt [hunk 13]
```diff
@@ -70,6 +79,45 @@
...
+*set cslib_ret ""
+*temp found_it false
+*temp n 1
+*temp array_count {p_object&"_count"}
+*temp index ""
+*label _get_by_field_main_loop
+*if (n <= array_count)
+	*if ({p_object&"_${n}_${p_field}"} = p_value)
+		*set found_it true
+		*goto end_search
```
Comment (CareyJWilliams) [inline scenes/cslib_object.txt:111]: _end_search

File: scenes/cslib_object.txt [hunk 14]
```diff
@@ -70,6 +79,45 @@
...
+	*set n + 1
+	*goto _get_by_field_main_loop
+*label end_search
+*if (found_it)
```
Comment (CareyJWilliams) [inline scenes/cslib_object.txt]:
```suggestion
*if (n <= array_count)
	*if ({p_object&"_${n}_${p_field}"} = p_value)
		*set found_it true
		*goto _get_by_field_end_search
	*set n + 1
	*goto _get_by_field_main_loop
*label _get_by_field_end_search
*if (found_it)
```

**Review by CareyJWilliams — APPROVED** [review]

**Comment by CareyJWilliams** [comment]
LGTM. Thank you Stephen!
